File: 0006-zigzag-conversion/0006-zigzag-conversion.java
class Solution {
    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        StringBuilder[] stacks = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) {
            stacks[i] = new StringBuilder();
        }
        for (int i = 0; i < s.length(); i++) {
            stacks[cycle(i, numRows)].append(s.charAt(i));
        }

        for (int i = 1; i < numRows; i++) {
            stacks[0].append(stacks[i]);
        }

        return stacks[0].toString();
    }

    int cycle(int i, int n) {
        int cycle = (n - 1) * 2;
        int pos = i % cycle;

        return pos < n ? pos : cycle - pos;
    }
}

File: 0006-zigzag-conversion/0006-zigzag-conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        if num_rows == 1 {
            return s;
        }

        let num_rows = num_rows as usize;
        let mut stacks = vec![String::new(); num_rows];

        for (i, ch) in s.chars().enumerate() {
            let row = Self::cycle(i, num_rows);
            stacks[row].push(ch);
        }

        stacks.concat()
    }

    fn cycle(i: usize, n: usize) -> usize {
        let cycle = (n - 1) * 2;
        let pos = i % cycle;

        if pos < n {
            pos
        } else {
            cycle - pos
        }
    }
}